### DISCUSSION TITLE
[v2] AppData App directory not being created

### DISCUSSION THREAD

**Discussion by Linkek** [tauri-apps/tauri#11279] [Q&A]
Hey everyone,

I have been trying to save some data in the AppData/Roaming/com.app-name.app folder. But the App directory is not being created. I'm not sure if it's bec of a bug or my config. But yea as the ".../Roaming/com.app-name.app" is none existing I cant create, save any files there. 

As I understood it, Tauri should just generate it for me. 

Here is my config:

```{
  "$schema": "https://schema.tauri.app/config/2.0.0-rc",
  "productName": "com.app-name.app",
  "version": "0.1.0",
  "identifier": "com.app-name.app",
  "build": {
    "beforeDevCommand": "yarn dev",
    "devUrl": "http://localhost:1420",
    "beforeBuildCommand": "yarn build",
    "frontendDist": "../dist"
  },
  "app": {
    "windows": [
      {
        "title": "app-name",
        "width": 800,
        "height": 600
      }
    ],
    "security": {
      "csp": null
    }
  },
  "bundle": {
    "active": true,
    "targets": "all",
    "icon": [
      "icons/32x32.png",
      "icons/128x128.png",
      "icons/128x128@2x.png",
      "icons/icon.icns",
      "icons/icon.ico"
    ]
  }
}
```

default.json

```{
  "$schema": "../gen/schemas/desktop-schema.json",
  "identifier": "default",
  "description": "Capability for the main window",
  "windows": ["main"],
	"scope:": ["$APPDATA/**"],
  "permissions": [
    "core:default",
    "shell:allow-open",
		"fs:allow-read-text-file",
		"fs:allow-write-text-file",
		"fs:allow-exists",
		"fs:allow-mkdir",
		{
      "identifier": "fs:scope",
      "allow": [
        { "path": "$APPDATA/**" }
      ]
    }
  ],
	"fs": {
    "scope": ["$APPDATA/**"],
    "allow": ["readTextFile", "writeTextFile", "mkdir", "exists"]
  }
}
```

Any ideas would be appreciated!

**Comment by FabianLars** [comment]
> As I understood it, Tauri should just generate it for me.

What made you think that? Did our docs hint at that? If so, then please tell us about it because we need to fix that then.

Generally, Tauri does not create any directories it doesn't need internally so you always want to (try to) create the directories you need before writing to them.

**Reply by stet** [reply]
What is the proper process to create app dir(s) then? Do they need to be created by tauri when in dev or can they be created manually for testing fs plugin etc?

**Reply by FabianLars** [reply]
You can create them by hand if you like. Otherwise tauri provides the mkdir createDir (v1) / mkdir (v2) methods for this on the js side. Rust has create_dir / create_dir_all for the rust side built-in.

**Reply by arashi-dev** [reply]
@FabianLars 
so we need to always repeat this code every time we want to write to it?

```ts
await createDir(await getIdentifier(), {
  baseDir: BaseDirectory.AppData,
  recursive: true,
});
```

and hardcode it in capabilities for every commands? 
```json
{
  "identifier": "fs:default",
  "allow": [
    {
      "path": "$APPDATA/com.app.name"
    }
 ]
},
...
```

isn't there any easier workaround? 
It would be much better to add another `BaseDirectory` that handles it automatically.

**Reply by FabianLars** [reply]
BaseDirectory.AppData  already contains the identifier.
In case this happened here, do not confuse it with the Windows specific `AppData` folder (the parent of Local, LocalLow, Roaming)

**Reply by arashi-dev** [reply]
Now I understand! thanks for clarification. but I think it still makes confusion (at least it did to me) that it should make that directory if the `BaseDirectory.AppData` is pointing to. otherwise what's the point if it is not solving anything! I should NOT make directories every time I need them if they are the **base** directories. if a new developer comes to project, they would think: "if Tauri is handling base paths, then it should also handle their filesystem, why not"

Having this code:
```ts
```ts
await write(filename, data, {
    baseDir: BaseDirectory.AppData,
    create: true,
});
```
makes much more sense, than having:
```ts
await mkdir(await appDataDir());

await write(filename, data, {
    baseDir: BaseDirectory.AppData,
    create: true,
});
```

Is there something I am missing?

**Reply by proxie-ghanshyam** [reply]
Hey, i faced the same problem recently. I think it should be mentioned in filesystem or path api like appdatadir also needs to be created. The problem accours that when i read dir name is "addDataDir". it feels like appDatadir is a app's installed dir which should exists and contains some app files. I also didnt knew that we have to manually create the appDataDir for the app too.

I solved it as mentioned above, the solution is pretty simple. I just misunderstood too.

**Reply by raphaelnunes67** [reply]
I had the same problem. I was using the file system to save some settings from my application, such as language and my app’s state. On Windows, I found that the Tauri application wasn’t able to create my app’s folder under C:\Users\<User>\AppData\Roaming. The workaround was to add handling in my “storage” class to save under C:\Users\<User>\AppData\Local.

```ts
import { 
  writeTextFile, 
  readTextFile, 
  rename, 
  exists, 
  BaseDirectory,
  remove,
  mkdir
} from '@tauri-apps/plugin-fs';

const PROFILES_FILE = 'profiles.json';
const SETTINGS_FILE = 'settings.json';

export class Storage {

  constructor(){}

  private getBaseDir(): BaseDirectory {
    const isWindows = navigator.userAgent.includes('Windows');
    return isWindows ? BaseDirectory.AppLocalData : BaseDirectory.AppData;
  }

  async init(): Promise<void> {
    const baseDir = this.getBaseDir();

    try {

      await mkdir('', { baseDir: baseDir, recursive: true });

      if (!await exists(SETTINGS_FILE, { baseDir: baseDir })) {
        await this.atomicWrite(SETTINGS_FILE, DEFAULT_SETTINGS);
      }
      if (!await exists(PROFILES_FILE, { baseDir: baseDir })) {
        await this.atomicWrite(PROFILES_FILE, []);
      }
    } catch (error) {
      console.error('Storage: Initialization failed', error);
    }
  }
  }
```

**Comment by SolarHez** [comment]
"scope": ["$APPDATA/**"] can be changed to "scope": ["**/**"] (this method is not secure but it works).
